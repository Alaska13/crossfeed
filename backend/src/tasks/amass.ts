import { connectToDatabase, Domain } from '../models';
import { spawnSync } from 'child_process';
import { readFileSync, unlinkSync } from 'fs';
import { plainToClass } from 'class-transformer';
import { CommandOptions } from './ecs-client';
import getRootDomains from './helpers/getRootDomains';
import saveDomainsToDb from './helpers/saveDomainsToDb';

const OUT_PATH = 'out-' + Math.random() + '.txt';

export default async (commandOptions: CommandOptions) => {
  const { organizationId, organizationName } = commandOptions;

  console.log('Running amass on organization', organizationName);

  const rootDomains = await getRootDomains(organizationId);

  for (const rootDomain of rootDomains) {
    const args = [
      'enum',
      '-ip',
      '-active',
      '-d',
      rootDomain,
      '-json',
      OUT_PATH
    ];
    console.log('Running amass with args', args);
    spawnSync('amass', args, { stdio: 'pipe' });

    const output = String(readFileSync(OUT_PATH));
    const lines = output.split('\n');
    const domains: Domain[] = [];
    for (const line of lines) {
      if (line == '') continue;
      const parsed = JSON.parse(line);
      domains.push(
        plainToClass(Domain, {
          ip: parsed.addresses[0].ip,
          name: parsed.name,
          asn: parsed.addresses[0].asn,
          organization: organizationId
        })
      );
    }
    await saveDomainsToDb(domains);
    console.log(`amass created/updated ${domains.length} new domains`);
    unlinkSync(OUT_PATH);
  }
};
