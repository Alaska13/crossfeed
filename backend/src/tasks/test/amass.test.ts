import { handler as amass } from '../amass';
jest.mock('../helpers/getRootDomains');
jest.mock('../helpers/saveDomainsToDb');

describe('amass', () => {
  test('basic test', async () => {
    await amass({
      organizationId: 'organizationId',
      organizationName: 'organizationName',
      scanId: 'scanId',
      scanName: 'scanName',
      scanTaskId: 'scanTaskId'
    });
  });
});
