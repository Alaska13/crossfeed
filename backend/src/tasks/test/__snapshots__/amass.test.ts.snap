// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`amass basic test: helpers.saveDomainsToDb 1`] = `
Array [
  Domain {
    "asn": 6762,
    "ip": "2a02:26f0:7a00:195::447a",
    "name": "filedrop.cisa.gov",
    "organization": Object {
      "id": "organizationId",
    },
  },
  Domain {
    "asn": 6762,
    "ip": "2a02:26f0:7a00:188::447a",
    "name": "www.filedrop.cisa.gov",
    "organization": Object {
      "id": "organizationId",
    },
  },
]
`;
